mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,651 @@
+
+        Returns:
+            A tuple of the updated action selection state and the chosen action.
+        """
+        params, hidden_state, t, key = action_selection_state
+
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        obs = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], obs)
+        term_or_trunc = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], term_or_trunc)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(
+            params, hidden_state, (obs, term_or_trunc), eps
+        )
+
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = action[0, ...]  # (1, B, A) -> (B, A)
+
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+
+        return next_action_selection_state, action
+
+    def action_step(action_state: ActionState, _: Any) -> Tuple[ActionState, Dict]:
+        """Selects action, steps env, stores timesteps in rb and repacks the parameters."""
+        # light unpacking
+        action_selection_state, env_state, buffer_state, obs, terminal, term_or_trunc = action_state
+
+        # select the actions to take
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, term_or_trunc
+        )
+
+        # step env with selected actions
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = next_timestep.reward
+
+        transition = Transition(
+            obs, action, reward, terminal, term_or_trunc, next_timestep.extras["real_next_obs"]
+        )
+        # Add dummy time dim
+        transition = jax.tree_util.tree_map(lambda x: x[:, jnp.newaxis, ...], transition)
+        next_buffer_state = rb.add(buffer_state, transition)
+
+        # Next obs and term_or_trunc for learner state
+        next_obs = next_timestep.observation
+        # make compatible with network input and transition storage in next step
+        next_terminal = (1 - next_timestep.discount[..., 0, jnp.newaxis]).astype(bool)
+        next_term_or_trunc = next_timestep.last()[..., jnp.newaxis]
+
+        # Repack
+        new_act_state = ActionState(
+            next_action_selection_state,
+            next_env_state,
+            next_buffer_state,
+            next_obs,
+            next_terminal,
+            next_term_or_trunc,
+        )
+
+        return new_act_state, next_timestep.extras["episode_metrics"]
+
+    # Training functions
+
+    def prep_inputs_to_scannedrnn(obs: Observation, term_or_trunc: chex.Array) -> chex.Array:
+        """
+        Prepares the inputs to the RNN network for either
+        getting q values or the eps-greedy distribution.
+        """
```

Comment:
nit: more detail here
```suggestion
        """Prepares the inputs to the RNN network for either getting q values or the 
        eps-greedy distribution.
        
        Mostly swaps leading axes because the replay buffer outputs (B, T, ... )
        and the RNN takes in (T, B, ...).
        """
```